content/src/content/jcr_root/apps/core/wcm/components/commons/datalayer/clientlibs/js/datalayer.js

```diff
@@ -0,0 +1,183 @@
+            updateState(window.dataLayer.state, event.object);
+        } else if (event.type === "listenerdefined") {
+            registerListener(event);
+        }
+    }
+
+    function registerListener(event) {
+        console.log("register an event listener: " + event);
+    }
+
+    function updateState(state, event) {
+        deepMerge(state, event.object);
+    }
+
+    function deepMerge(target, source) {
+        var tmpSource = {};
+        if (isObject(target) && isObject(source)) {
+            Object.keys(source).forEach(function(key) {
+                if (isObject(source[key])) {
+                    if (!target[key]) {
+                        tmpSource[key] = {};
+                        Object.assign(target, tmpSource);
+                    }
+                    deepMerge(target[key], source[key]);
+                } else {
+                    tmpSource[key] = source[key];
+                    Object.assign(target, tmpSource);
+                }
+            });
+        }
+    }
+
+    function isObject(item) {
+        return (item && typeof item === "object" && !Array.isArray(item));
+    }
+
+    function populateDataLayer() {
+        window.dataLayer.push({
+            "type": "updated",
+            "object": {
+                "component": {
+                    "carousel": {
+                        "carousel3": {
+                            "id": "/content/mysite/en/home/jcr:content/root/carousel3",
+                            "items": {}
+                        }
+                    }
+                }
+            }
+        });
+
+        window.dataLayer.push({
+            "type": "updated",
+            "object": {
+                "component": {
+                    "tab": {
+                        "tab2": {
+                            "id": "/content/mysite/en/home/jcr:content/root/tab2",
+                            "items": {}
+                        }
+                    }
+                }
+            }
+        });
+
+        window.dataLayer.push({
+            "type": "updated",
+            "object": {
+                "page": {
+                    "id": "/content/mysite/en/products/crossfit",
+                    "siteLanguage": "en-us",
+                    "siteCountry": "US",
+                    "pageType": "product detail",
+                    "pageName": "pdp - crossfit zoom",
+                    "pageCategory": "womens > shoes > athletic"
+                }
+            }
+        });
+
+        window.dataLayer.push({
+            "type": "listenerdefined",
+            "object": {
+                "listener": {
+                    "type": "clicked",
+                    "target": {},
+                    "handler": function(component) {
+                        console.log("clicked on component: " + component);
+                    }
+                }
+            }
+        });
+    }
+
+    function populateDataLayerAfterOverride() {
```

Comment:
for now, we can trigger a custom event on the window object once we've initialized:
```
var readyEvent = new CustomEvent("datalayer:ready");
this.window.dispatchEvent(readyEvent);
```

and add the after population based on that:
```
window.addEventListener("datalayer:ready", function() { console.log("data layer ready"); });
```

will allow us to decouple the mocks you have already from the core handling.